lib/pure/strutils.nim

```diff
@@ -979,6 +992,46 @@ proc parseHexInt*(s: string): int {.noSideEffect, procvar,
     of '\0': break
     else: raise newException(ValueError, "invalid integer: " & s)
 
+proc generateHexCharToValueMap(): string =
+  ## Generate a string to map a hex digit to uint value
+  result = ""
+  for inp in 0..255:
+    let ch = chr(inp)
+    let o =
+      case ch:
+        of '0'..'9': ord('0')
+        of 'a'..'f': ord('a') - 10
+        of 'A'..'F': ord('A') - 10
+        else: -1
+    if o == -1:
+      result.add chr(17) # indicates an invalid hex char
+    else:
+      result.add chr(inp - o)
+
+const hex_char_to_value_map = generateHexCharToValueMap()
+
+proc parseHexStr*(s: string): string {.noSideEffect, procvar,
+  rtl, extern: "nsuParseHexStr".} =
+  ## Convert hex-encoded string to byte string, e.g.:
+  ##
+  ## .. code-block:: nim
+  ##    hexToStr("00ff") == "\0\255"
+  ##
+  ## Raises ``ValueError`` for an invalid hex values. The comparison is
+  ## case-insensitive.
+  if s.len mod 2 != 0:
+    raise newException(ValueError, "Incorrect hex string len")
+  result = newString(s.len div 2)
+  var buf = 0
+  for pos, c in s:
+    let val = hex_char_to_value_map[ord(c)].ord
```

Comment:
Do we really need this lookup table solution here? The table is full of chr(17).

Reply:
I ran nimbench, the lookup is faster than alternatives like doing comparisons and subtractions at runtime.